RPC to trace execution of specified block

mattrutherford opened the pull request:
Provide new RPC: `state_traceBlock`

- Returns all `Span`s and `Event`s while executing the historic block.
- Allows overriding targets with defaults set to pallet, frame and state

eg.

```
curl -H "Content-Type: application/json" -d '{"id":1, "jsonrpc":"2.0", "method": "state_traceBlock", "params": ["0x5ae2a90b69935917b28546b195b307876cfc3561f794fbd62a1464abd4c799a5"]}' http://localhost:9933/
```

Docs can be found in _client/rpc-api/src/state/mod.rs_

mattrutherford commented:
Example response (formatted): https://gist.github.com/mattrutherford/6fa415abc76de62763daed88eaba2b25

insipx commented:
Maybe we could avoid serializing `bool/i64` values if they're empty to leave them out of the JSON, that might de clutter things a bit. Otherwise, I think it's OK to leave each trace as it's own isolated JSON Object. The only other thing that comes to mind is maybe moving block hash/block number to the top-level so that a user could avoid manually looking for the span that defines the block that is being executed.

dvdplm reviewed (commented):
I'll be back for a deeper review, left some questions.

Is the `Mutex` in `StorageTracer` a cause for concern you think? I guess executing a single block at a time is not particularly performance critical, but if there are many events emitted I could see us ending up spending a lot of time spinning on the lock.

The PR description could use some love: intended use case, how to compile&run this etc.

emostov commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,132 @@
+				target: TRACE_TARGET,
+				"trace_block",
+				block_hash = ?self.block,
+				?parent_hash,
```
Not sure if I am reading this correctly, but what does the leading `?` do? I am used to seeing it trailing for unwrap or error, but google is not telling me what a leading one does.

mattrutherford replied:
it's a DSL specific to the `tracing` macro - it basically means shorthand to use the `Debug` representation of the variable.

emostov commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,132 @@
+		let id = BlockId::Hash(self.block);
+		let extrinsics = self.client.block_body(&id)
+			.map_err(|e| format!("Invalid block id: {:?}", e))?
+			.ok_or("Block not found".to_string())?;
```
Should this be "Extrinsics not found"?

mattrutherford replied:
yes, maybe, but I should probably change the order to get the header first, in case it's just that the block doesn't exist then the error would be a bit misleading

emostov commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,132 @@
+			return Err(format!("Error executing block: {:?}", e));
+		}
+
+		drop(dispatch);
```
Is there something I can read to explain how `drop` is being used here? Looking in `tracing` crate docks but don't see a mention of a special `drop` function. I also don't understand where this function is defined or imported.

mattrutherford replied:
this actually just the `std::mem::drop` function - it's just a way to drop the dispatch early before it goes out of scope so that we can unwrap the `Arc` etc and return an owned `Traces`

emostov reviewed (commented):
I left a few trivial comments, but I am going to need to spend a good chunk of time with this tomorrow to (attempt to) give a review with some integrity

mattrutherford commented on client/rpc-api/src/state/mod.rs:
```diff
@@ -136,4 +136,10 @@ pub trait StateApi<Hash> {
 		&self, metadata: Option<Self::Metadata>, id: SubscriptionId
 	) -> RpcResult<bool>;
+
+	/// Execute block and return all storage traces
```
Actually we don't need any special flag because we set up a dispatch as part of this. WRT wording, I think you're right it could be improved and I need to add some more details - will push a commit to address it.

mattrutherford commented:
> I'll be back for a deeper review, left some questions.
> 
> Is the `Mutex` in `StorageTracer` a cause for concern you think? I guess executing a single block at a time is not particularly performance critical, but if there are many events emitted I could see us ending up spending a lot of time spinning on the lock.
> 
> The PR description could use some love: intended use case, how to compile&run this etc.

Good point I will update the description. I think the Mutex is fine - the lock wont be contested because the runtime is single-threaded and I think any alternative would be too complex.

mattrutherford commented on client/tracing/src/lib.rs:
```diff
@@ -417,14 +456,14 @@ impl<S: Subscriber> Layer<S> for ProfilingLayer {
 		};
 
 		if let Some(mut span_datum) = span_datum {
+			self.current_span.exit();
```
Because if it was none - we'd be exiting the wrong span (next one up in the hierarchy).

mattrutherford commented on client/tracing/src/lib.rs:
```diff
@@ -417,14 +456,14 @@ impl<S: Subscriber> Layer<S> for ProfilingLayer {
 		let end_time = Instant::now();
 		let span_datum = {
 			let mut span_data = self.span_data.lock();
 			span_data.remove(&span)
```
It shouldn't - alternative is probably to make it an `expect`

mattrutherford commented on client/tracing/src/lib.rs:
```diff
@@ -148,7 +151,7 @@ pub fn parse_default_directive(directive: &str) -> Result<Directive, String> {
 	targets: Vec<(String, Level)>,
 	trace_handler: Box<dyn TraceHandler>,
-	span_data: Mutex<FxHashMap<Id, SpanDatum>>,
+	span_data: Mutex<HashMap<Id, SpanDatum>>,
```
I had some ergonomic problems, but actually I don't remember exactly what - I will try to add it back

emostov replied:
Any benefit in adding it back?

insipx commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,132 @@
+			.ok_or("Block not found".to_string())?;
+		let parent_hash = header.parent_hash().clone();
+		let parent_id = BlockId::Hash(parent_hash.clone());
+		// Pop digest else RuntimePanic due to: 'Number of digest items must match that calculated.'
```
iirc once a block is finalized, a digest item is added post-execution based on consensus of the chain. Because of that, re-executing a block will fail because of this extra digest item since the runtime doesn't expect it

mattrutherford commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,132 @@
+	pub fn trace_block(&self) -> Result<Traces, String> {
+		// Prepare block
+		let id = BlockId::Hash(self.block);
+		let extrinsics = self.client.block_body(&id)
```
that was my thinking, also `Block::new` takes `header, extrinsics` as arguments, so it matches with that

dvdplm replied:
Cool, I stand corrected.

mattrutherford commented:
> Maybe we could avoid serializing `bool/i64` values if they're empty to leave them out of the JSON, that might de clutter things a bit. Otherwise, I think it's OK to leave each trace as it's own isolated JSON Object. The only other thing that comes to mind is maybe moving block hash/block number to the top-level so that a user could avoid manually looking for the span that defines the block that is being executed.

Seems like a good idea to move block hash to top level. 

I'm not sure about skipping empty values, I can definitely see the advantage for sure, seeing as we're not actually using any yet, but wondering if it still better to include them in case they do get used at some point - ie. is it better to be explicit/consistent (seeing as we don't know how this will be consumed)?

emostov reviewed (commented):
Overall I think it looks good. Just some notes:
- My understanding is a bit shaky on exactly where the hooks for the profiling layer are called, but their logic seems reasonable. 
- I can't really evaluate the use of shared resources (Mutex) since I don't have experience thinking about that.
- It is not clear to me how the `message` portion of the traces will be able to be processed by the the user of the RPC, and if the assumptions made by that user will need to change with runtime upgrades
- How do we deal with conceptual level docs for tracing? I would think explaining targets and the structure of data would be helpful

dvdplm commented on client/rpc-api/src/state/mod.rs:
```diff
@@ -136,4 +136,10 @@ pub trait StateApi<Hash> {
+	/// Execute block and return all storage traces
+	#[rpc(name = "state_traceBlock")]
+	fn trace_block(
+		&self, block: Hash, targets: Option<String>
```
I agree. It's one of those things that aren't obvious until you learn it and then you never think about it again. The string here is in the Rust logging target format, equal to the stuff you'd pass in on the command line when starting a node. Examples:
`debug`: log all targets at debug, warn or error levels.
`rpc=debug`: leave default levels for all targets except `rpc` that logs at `debug` level.
`off,grandpa=trace`: turn off all logging except the `grandpa` target that logs at `trace` levels
`info,rpc=trace,foo=warn`: …well, you get it. :)

bkchr commented on client/rpc-api/src/state/mod.rs:
```diff
@@ -136,4 +136,10 @@ pub trait StateApi<Hash> {
+	/// Execute block and return all storage traces
+	#[rpc(name = "state_traceBlock")]
+	fn trace_block(
+		&self, block: Hash, targets: Option<String>
```
```suggestion
		&self,
		block: Hash,
		targets: Option<String>,
```

bkchr commented on client/rpc/src/state/mod.rs:
```diff
@@ -165,6 +167,12 @@ pub trait StateBackend<Block: BlockT, Client>: Send + Sync + 'static
+	/// Trace storage changes for block
+	fn trace_block(
+		&self,
+		block: Block::Hash, targets: Option<String>
```
```suggestion
		block: Block::Hash,
		targets: Option<String>,
```

bkchr commented on client/rpc/src/state/state_full.rs:
```diff
@@ -521,12 +526,23 @@ impl<BE, Block, Client> StateBackend<Block, Client> for FullState<BE, Block, Cli
+		&self,
+		block: Block::Hash, targets: Option<String>,
+	) -> FutureResult<sp_rpc::tracing::BlockTrace> {
+		Box::new(result(
```
This should be an unsafe RPC that is only enabled when unsafe RPC are enabled.

bkchr commented on client/rpc/src/state/state_light.rs:
```diff
@@ -474,6 +474,13 @@ impl<Block, F, Client> StateBackend<Block, Client> for LightState<Block, F, Clie
+
+	fn trace_block(
+		&self,
+		_block: Block::Hash, _targets: Option<String>
```
```suggestion
		_block: Block::Hash,
		_targets: Option<String>,
```

bkchr commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,253 @@
+// Copyright 2019-2020 Parity Technologies (UK) Ltd.
```
```suggestion
// Copyright 2021 Parity Technologies (UK) Ltd.
```

bkchr commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,253 @@
+
+//! Utilities for tracing block execution
+
+use std::collections::HashMap;
+use std::sync::{Arc};
+use std::sync::atomic::{AtomicU64, Ordering};
```
```suggestion
use std::{collections::HashMap, sync::{Arc, atomic::{AtomicU64, Ordering}}};
```

bkchr commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,253 @@
+
+use tracing_core::span::{Attributes, Record, Id};
+use tracing_core::{Level};
+use std::time::{Instant};
```
```suggestion
```

please use `wasm_timer::Instant`

bkchr commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,253 @@
+		if let Err(e) = dispatcher::with_default(&dispatch, || {
+			let span = tracing::info_span!(
+				target: TRACE_TARGET,
+				"trace_block"
```
```suggestion
				"trace_block",
```

bkchr commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,253 @@
+	targets: Vec<(String, Level)>,
+	next_id: AtomicU64,
+	current_span: CurrentSpan,
+	spans: Arc<Mutex<HashMap<Id, Span>>>,
+	events: Arc<Mutex<Vec<Event>>>,
```
Why are these `Arc`s?

dvdplm replied:
`tracing::Dispatch` takes an owned `Subscriber` with constraints `S: Subscriber + Send + Sync + 'static, ` so `Rc` doesn't work.

The [way we capture the traces from a single block](https://github.com/paritytech/substrate/pull/7780/files#diff-f5ee9da45314c57567317ca337c1686d02bcf4d13779775ec99186d275e51957R196-R206) is through a new local Dispatcher within which we execute the block; then we drop the `Dispatcher` and unwrap the `Arc` to get the spans and events.

I don't see a better way, do you?

bkchr replied:
Ahh I had overseen this.

There would another way by using this: https://docs.rs/tracing/0.1.22/tracing/struct.Dispatch.html#method.downcast_ref

dvdplm replied:
That's a good idea; I tried this [here](https://github.com/paritytech/substrate/pull/7780/commits/d9fd09d33cfcffe005c9c82d37a0668e04844891) but there's [a test](https://github.com/paritytech/substrate/blob/5c183f3b1110e1cefdc74a2f8787187d6080ffa8/client/tracing/src/lib.rs#L630) that is behaving weirdly with this code. Not sure if it's the code or the test (or both).
`cargo t -p sc-tracing` *skips* the test for some reason, but filtering for it reveals the failure (`cargo t -p sc-tracing thread`).
TBH I'm not sure if the test is correct or not.

bkchr replied:
This test is failing or what? :D

dvdplm replied:
It is failing. But as you see all tests pass in CI here, so wtf? Turns out that the tests in `logging/`, e.g. [here](https://github.com/paritytech/substrate/blob/5c183f3b1110e1cefdc74a2f8787187d6080ffa8/client/tracing/src/logging/mod.rs#L389-L394) that execute out-of-process somehow swallow the other tests; I'm guessing it messes up `stderr`. 
`cargo t -p sc-tracing` claims to run 14 tests, but actually just runs 8. Sometimes 9 (the one failing), and when it does, the test fails but does not report the failure to the test runner, so it's like it was never there.

So I have two problems:
1. Fix the out-of-process tests (I'll try moving them to a crate and see if that helps; cc @cecton)
2. Figure out if the failing tracing test should fail or pass. A question I have here is if, in your opinion, that test is correct or not.

bkchr commented on primitives/io/src/lib.rs:
```diff
@@ -86,7 +86,13 @@ pub enum EcdsaVerifyError {
 	fn get(&self, key: &[u8]) -> Option<Vec<u8>> {
-		self.storage(key).map(|s| s.to_vec())
+		let res = self.storage(key).map(|s| s.to_vec());
+		sp_tracing::debug!(
```
We had this already in a previous pr of you, I don't see why you add this?

This is already traced by `Ext`.

So please revert all the changes in this file and the `Cargo.toml`.

bkchr commented on primitives/state-machine/src/ext.rs:
```diff
@@ -222,10 +222,11 @@ where
-			self.id,
-			HexDisplay::from(&key),
-			result.as_ref().map(HexDisplay::from)
+		trace!(target: "state",
+			method = "Get",
+			ext_id = self.id,
+			key = %HexDisplay::from(&key),
+			result = ?result.as_ref().map(HexDisplay::from)
 		);
```
```suggestion
		trace!(
			target: "state",
			method = "Get",
			ext_id = self.id,
			key = %HexDisplay::from(&key),
			result = ?result.as_ref().map(HexDisplay::from),
		);
```

bkchr commented on primitives/state-machine/src/ext.rs:
```diff
@@ -237,10 +238,11 @@ where
-			self.id,
-			HexDisplay::from(&key),
-			result,
+		trace!(target: "state",
+			method = "Hash",
+			ext_id = self.id,
+			key = %HexDisplay::from(&key),
+			result = ?result
 		);
```
```suggestion
		trace!(
			target: "state",
			method = "Hash",
			ext_id = self.id,
			key = %HexDisplay::from(&key),
			result = ?result,
		);
```

bkchr commented on primitives/state-machine/src/ext.rs:
```diff
@@ -259,11 +261,12 @@ where
-			HexDisplay::from(&child_info.storage_key()),
-			HexDisplay::from(&key),
-			result.as_ref().map(HexDisplay::from)
+		trace!(target: "state",
+			method = "GetChild",
+			ext_id = self.id,
+			child_key = %HexDisplay::from(&child_info.storage_key()),
+			key = %HexDisplay::from(&key),
+			result = ?result.as_ref().map(HexDisplay::from)
 		);
```
```suggestion
		trace!(
			target: "state",
			method = "GetChild",
			ext_id = self.id,
			child_key = %HexDisplay::from(&child_info.storage_key()),
			key = %HexDisplay::from(&key),
			result = ?result.as_ref().map(HexDisplay::from),
		);
```

bkchr commented on primitives/state-machine/src/ext.rs:
```diff
@@ -283,11 +286,12 @@ where
-			HexDisplay::from(&child_info.storage_key()),
-			HexDisplay::from(&key),
-			result,
+		trace!(target: "state",
+			method = "ChildHash",
+			ext_id = self.id,
+			child_key = %HexDisplay::from(&child_info.storage_key()),
+			key = %HexDisplay::from(&key),
+			result = ?result
 		);
```
```suggestion
		trace!(
			target: "state",
			method = "ChildHash",
			ext_id = self.id,
			child_key = %HexDisplay::from(&child_info.storage_key()),
			key = %HexDisplay::from(&key),
			?result,
		);
```

bkchr commented on primitives/state-machine/src/ext.rs:
```diff
@@ -300,10 +304,11 @@ where
-			self.id,
-			HexDisplay::from(&key),
-			result,
+		trace!(target: "state",
+			method = "Exists",
+			ext_id = self.id,
+			key = %HexDisplay::from(&key),
+			result = ?result
 		);
```
```suggestion
		trace!(
			target: "state",
			method = "Exists",
			ext_id = self.id,
			key = %HexDisplay::from(&key),
			?result,
		);
```

bkchr commented on primitives/state-machine/src/ext.rs:
```diff
@@ -323,11 +328,12 @@ where
-			HexDisplay::from(&child_info.storage_key()),
-			HexDisplay::from(&key),
-			result,
+		trace!(target: "state",
+			method = "ChildExists",
+			ext_id = self.id,
+			child_key = %HexDisplay::from(&child_info.storage_key()),
+			key = %HexDisplay::from(&key),
+			result = ?result
 		);
```
```suggestion
		trace!(
			target: "state",
			method = "ChildExists",
			ext_id = self.id,
			child_key = %HexDisplay::from(&child_info.storage_key()),
			key = %HexDisplay::from(&key),
			?result,
		);
```

bkchr commented on primitives/state-machine/src/ext.rs:
```diff
@@ -375,10 +381,11 @@ where
-			self.id,
-			HexDisplay::from(&key),
-			value.as_ref().map(HexDisplay::from)
+		trace!(target: "state",
+			method = "Put",
+			ext_id = self.id,
+			key = %HexDisplay::from(&key),
+			value = ?value.as_ref().map(HexDisplay::from)
 		);
```
```suggestion
		trace!(
			target: "state",
			method = "Put",
			ext_id = self.id,
			key = %HexDisplay::from(&key),
			value = ?value.as_ref().map(HexDisplay::from),
		);
```

bkchr commented on primitives/state-machine/src/ext.rs:
```diff
@@ -396,11 +403,12 @@ where
-			HexDisplay::from(&child_info.storage_key()),
-			HexDisplay::from(&key),
-			value.as_ref().map(HexDisplay::from)
+		trace!(target: "state",
+			method = "PutChild",
+			ext_id = self.id,
+			child_key = %HexDisplay::from(&child_info.storage_key()),
+			key = %HexDisplay::from(&key),
+			value = ?value.as_ref().map(HexDisplay::from)
 		);
```
```suggestion
		trace!(
			target: "state",
			method = "PutChild",
			ext_id = self.id,
			child_key = %HexDisplay::from(&child_info.storage_key()),
			key = %HexDisplay::from(&key),
			value = ?value.as_ref().map(HexDisplay::from),
		);
```

bkchr commented on primitives/state-machine/src/ext.rs:
```diff
@@ -413,9 +421,10 @@ where
-		trace!(target: "state", "{:04x}: KillChild({})",
-			self.id,
-			HexDisplay::from(&child_info.storage_key()),
+		trace!(target: "state",
+			method = "KillChild",
+			ext_id = self.id,
+			child_key = %HexDisplay::from(&child_info.storage_key())
 		);
```
```suggestion
		trace!(
			target: "state",
			method = "KillChild",
			ext_id = self.id,
			child_key = %HexDisplay::from(&child_info.storage_key()),
		);
```

bkchr commented on primitives/state-machine/src/ext.rs:
```diff
@@ -449,9 +458,10 @@ where
-		trace!(target: "state", "{:04x}: ClearPrefix {}",
-			self.id,
-			HexDisplay::from(&prefix),
+		trace!(target: "state",
+			method = "ClearPrefix",
+			ext_id = self.id,
+			prefix = %HexDisplay::from(&prefix)
 		);
```
```suggestion
		trace!(
			target: "state",
			method = "ClearPrefix",
			ext_id = self.id,
			prefix = %HexDisplay::from(&prefix),
		);
```

bkchr commented on primitives/state-machine/src/ext.rs:
```diff
@@ -472,10 +482,11 @@ where
-			self.id,
-			HexDisplay::from(&child_info.storage_key()),
-			HexDisplay::from(&prefix),
+		trace!(target: "state",
+			method = "ClearChildPrefix",
+			ext_id = self.id,
+			child_key = %HexDisplay::from(&child_info.storage_key()),
+			prefix = %HexDisplay::from(&prefix)
 		);
```
```suggestion
		trace!(
			target: "state",
			method = "ClearChildPrefix",
			ext_id = self.id,
			child_key = %HexDisplay::from(&child_info.storage_key()),
			prefix = %HexDisplay::from(&prefix),
		);
```

bkchr commented on primitives/state-machine/src/ext.rs:
```diff
@@ -491,10 +502,11 @@ where
-			self.id,
-			HexDisplay::from(&key),
-			HexDisplay::from(&value),
+		trace!(target: "state",
+			method = "Append",
+			ext_id = self.id,
+			key = %HexDisplay::from(&key),
+			value = %HexDisplay::from(&value),
 		);
```
```suggestion
		trace!(
			target: "state",
			method = "Append",
			ext_id = self.id,
			key = %HexDisplay::from(&key),
			value = %HexDisplay::from(&value),
		);
```

bkchr commented on primitives/state-machine/src/ext.rs:
```diff
@@ -511,15 +523,20 @@ where
-			trace!(target: "state", "{:04x}: Root(cached) {}",
-				self.id,
-				HexDisplay::from(&root.as_ref()),
+			trace!(target: "state",
+				method = "Root(cached)",
+				ext_id = self.id,
+				root = %HexDisplay::from(&root.as_ref())
 			);
```
```suggestion
			trace!(
				target: "state",
				method = "Root",
				ext_id = self.id,
				root = %HexDisplay::from(&root.as_ref()),
				cached = true,
			);
```

bkchr commented on primitives/state-machine/src/ext.rs:
```diff
@@ -511,15 +523,20 @@ where
 
 		let root = self.overlay.storage_root(self.backend, self.storage_transaction_cache);
-		trace!(target: "state", "{:04x}: Root {}", self.id, HexDisplay::from(&root.as_ref()));
+		trace!(target: "state",
+			   method = "Root",
+			   ext_id = self.id,
+			   root = %HexDisplay::from(&root.as_ref())
+		);
```
```suggestion
		trace!(
			target: "state",
			method = "Root",
			ext_id = self.id,
			root = %HexDisplay::from(&root.as_ref()),
			cached = false,
		);
```

bkchr commented on primitives/state-machine/src/ext.rs:
```diff
@@ -537,10 +554,11 @@ where
-				self.id,
-				HexDisplay::from(&storage_key),
-				HexDisplay::from(&root.as_ref()),
+			trace!(target: "state",
+				method = "ChildRoot(cached)",
+				ext_id = self.id,
+				storage_key = %HexDisplay::from(&storage_key),
+				root = %HexDisplay::from(&root.as_ref()),
 			);
```
```suggestion
			trace!(
				target: "state",
				method = "ChildRoot",
				ext_id = self.id,
				storage_key = %HexDisplay::from(&storage_key),
				root = %HexDisplay::from(&root.as_ref()),
				cached = true,
			);
```

bkchr commented on primitives/state-machine/src/ext.rs:
```diff
@@ -564,10 +582,11 @@ where
-					self.id,
-					HexDisplay::from(&storage_key.as_ref()),
-					HexDisplay::from(&root.as_ref()),
+				trace!(target: "state",
+					method = "ChildRoot",
+					ext_id = self.id,
+					storage_key = %HexDisplay::from(&storage_key.as_ref()),
+					root = %HexDisplay::from(&root.as_ref()),
 				);
```
```suggestion
				trace!(
					target: "state",
					method = "ChildRoot",
					ext_id = self.id,
					storage_key = %HexDisplay::from(&storage_key.as_ref()),
					root = %HexDisplay::from(&root.as_ref()),
					cached = true,
				);
```

bkchr commented on primitives/state-machine/src/ext.rs:
```diff
@@ -578,10 +597,11 @@ where
-					self.id,
-					HexDisplay::from(&storage_key.as_ref()),
-					HexDisplay::from(&root.as_ref()),
+				trace!(target: "state",
+					method = "ChildRoot(no_change)",
+					ext_id = self.id,
+					storage_key = %HexDisplay::from(&storage_key.as_ref()),
+					root = %HexDisplay::from(&root.as_ref()),
 				);
```
```suggestion
				trace!(
					target: "state",
					method = "ChildRoot",
					ext_id = self.id,
					storage_key = %HexDisplay::from(&storage_key.as_ref()),
					root = %HexDisplay::from(&root.as_ref()),
				);
```

bkchr commented on primitives/state-machine/src/ext.rs:
```diff
@@ -610,10 +630,11 @@ where
-			self.id,
-			HexDisplay::from(&parent_hash),
-			root,
+		trace!(target: "state",
+			method = "ChangesRoot",
+			ext_id = self.id,
+			parent_hash = %HexDisplay::from(&parent_hash),
+			?root,
 		);
```
```suggestion
		trace!(
			target: "state",
			method = "ChangesRoot",
			ext_id = self.id,
			parent_hash = %HexDisplay::from(&parent_hash),
			?root,
		);
```

dvdplm commented on client/tracing/src/lib.rs:
```diff
@@ -29,8 +29,9 @@
+pub mod block;
 
-use rustc_hash::FxHashMap;
+// use rustc_hash::{FxHasher};
```
```suggestion
```

dvdplm commented on client/rpc/src/state/mod.rs:
```diff
@@ -347,6 +357,19 @@ impl<Block, Client> StateApi<Block::Hash> for State<Block, Client>
+	fn trace_block(&self, block: Block::Hash, targets: Option<String>) -> FutureResult<sp_rpc::tracing::BlockTrace> {
+		if let Err(err) = self.deny_unsafe.check_if_safe() {
+			// TODO: there has to be a more elegant way to do this.
+			return Box::new(async move { Err(err.into()) }.boxed().compat());
```
@bkchr this is terrible; what can I do better here? :/

It's a bit odd that different RPCs use different return types, for `system` there's `bail_if_unsafe!` but that doesn't work here.

bkchr replied:
Something like:
`return Box::new(rpc::futures::future::result(Err(err))`?

dvdplm replied:
ofc, ty very much.

mattrutherford commented:
Have reverted both the new events in storage and also the modifications in state-machine. If we need additional traces, these can be added in a follow-up PR, once the requirements are more clear regarding the format of data, values required and granularity.

I think for now, the main thing to settle in this PR is the format of the response.

@emostov  @dvdplm @insipx 

I have pushed 2 new branches with the extracted changes from this PR, which can be used as basis for new PRs if required:

`mr-storage-traces`
`mr-state-traces`

bkchr commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,251 @@
+			return Err(format!("Error executing block: {:?}", e));
+		}
+		drop(dispatch);
+		let spans = Arc::try_unwrap(spans)
+			.map_err(|_| "Unable to unwrap spans".to_string())?;
+		let events = Arc::try_unwrap(events)
+			.map_err(|_| "Unable to unwrap spans".to_string())?;
+		let spans: Vec<Span> = spans.into_inner().into_iter().map(|(_, s)| s.into()).collect();
+		// First filter out any spans that never entered
+		let spans: Vec<Span> = spans.into_iter().filter(|s| !s.entered.is_empty()).collect();
+		// Patch wasm identifiers
+		let mut spans: Vec<Span> = spans.into_iter().filter_map(|s| patch_and_filter(s, targets)).collect();
```
```suggestion
		let spans: Vec<Span> = spans.lock().drain().map(|(_, s)| s.into()).into_iter().filter(|s| !s.entered.is_empty()).filter_map(|s| patch_and_filter(s, targets)).collect();
```

You should get it ;) This way we will not have to do 3 collects, only one.

dvdplm commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,132 @@
+use crate::{TraceHandler, ProfilingLayer};
+
+// Default to only runtime and state related traces
+const DEFAULT_TARGETS: &'static str = "pallet,frame,state";
```
While I agree that having better docs around this is probably very useful to newcomers I also think that it's a lot of work to maintain it and that we likely do not have a fixed set of rules about when a new target should be added (and we certainly don't have a way to enforce any such rules), so in the end it comes down to single devs doing the right thing, which is a terrible base to build docs on. The gist of it is essentially that it behaves like "any Rust project", which is a terrible answer of course.
I don't mind extending the docs here with anything you think would be useful to readers.

bkchr commented on .maintain/monitoring/grafana-dashboards/README_dashboard.md:
```diff
@@ -5,3 +5,10 @@ Shared templated Grafana dashboards.
 documentation](https://grafana.com/docs/grafana/latest/reference/export_import/).
 You can see an example setup [here](../../../.maintain/sentry-node).
+
+#### Required labels on Prometheus metrics
```
This was deleted in master :D 

And the file below as well :D

dvdplm replied:
oh shit, it's all a mess now. :(

cecton replied:
That is why I prefer to resolve conflicting commits from master one by one :grin: https://github.com/cecton/git-tools#git-try-merge

emostov commented:
Still very much a WIP but I added in structured logging, which I think is the same goal as #8027.

dvdplm commented on primitives/state-machine/src/ext.rs:
```diff
@@ -194,18 +194,28 @@ where
 {
 	fn set_offchain_storage(&mut self, key: &[u8], value: Option<&[u8]>) {
+		// TODO REVIEW should we be tracing off chain storage? I don't think I need it for
+		// balance reconciliation in polkadot runtimes, but maybe someone else does?
```
No. We can add it when we need it.

emostov commented:
Considering the fact that we didn't catch the traces for hooks where missing in the new pallet attribute macro (fixed by #8305) I think we should consider some tests for the output of this RPC. rel: #8316

mattrutherford commented:
Added logging for wasm overrides, eg:

```
2021-04-09 14:39:26.250  INFO main wasm_overrides: Found wasm override in file: `Some("../substrate-archive/wasm-tracing/kusama/kusama_runtime_tracing_0.8.25.compact.wasm")`, version: kusama-2025 (parity-kusama-0.tx3.au2)
2021-04-09 14:39:27.270  INFO main wasm_overrides: Found wasm override in file: `Some("../substrate-archive/wasm-tracing/kusama/kusama_runtime_tracing_0.8.27.compact.wasm")`, version: kusama-2027 (parity-kusama-0.tx3.au2)
```

Added logging for state tracing:

```
2021-04-09 14:40:07.677  INFO        http.worker00 state_tracing: Tracing block: 0x45a0…5334
2021-04-09 14:40:12.874 DEBUG        http.worker00 sc_tracing: state_tracing: execute_block, time: 5194766697, id: 549755813889, parent_id: None, values: extrinsics_len=1
2021-04-09 14:40:12.874 DEBUG        http.worker00 state_tracing: close time.busy=5.19s time.idle=58.3µs
```

Also as part of this added feature to allow logging of spans in substrate, including the timings, as shown above.

mattrutherford commented on primitives/rpc/src/tracing.rs:
```diff
@@ -88,10 +88,13 @@ pub struct Span {
 #[derive(Serialize, Deserialize, Default, Clone, Debug)]
 pub struct Values {
 	/// HashMap of `bool` values
+	#[serde(skip)]
```
I think we should only skip these fields if empty - eg, we currently use a bool to indicate a wasm (vs native) trace.

emostov replied:
These are only used for serializing the JSON response, correct?

If so, I think we should leave it out for now since I don't think it is particularly helpful to have in the RPC response. (We can use other logging for that?). Seems like extra logic we can always add in later.

emostov replied:
@mattrutherford do we ever need to `Deserialize` with serde? Maybe we can remove that derive?

mattrutherford replied:
no we'll never need to, but afaicr it's a requirement of the RPC setup

mattrutherford replied:
> These are only used for serializing the JSON response, correct?
> 
> If so, I think we should leave it out for now since I don't think it is particularly helpful to have in the RPC response. (We can use other logging for that?). Seems like extra logic we can always add in later.

I think it could be helpful to know if a trace is WASM or native, because there's often no other way to confirm that from the name or target

emostov commented:
TODO:

- [ ] Filter events by key
- [ ] Remove traces from `primitives/state-machine/src/ext.rs` that are not `get` or `put`
- [ ] Allow key filters to be passed in as RPC param
- [ ] Put a max limit on events and spans (should limit apply after filter?)
- [ ] Custom `Values` `Serialization` impl that 
     - removes `"params": ""` field since it shows up empty on almost every span
     - remove `bool_values` if no `wasm` field
     - remove both `i64_values` and `u64_values` (this is already done just with `skip_serializing`, but should be included in a custom impl for the whole struct if we decide to go that route)

emostov commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,281 @@
+		events: Mutex<Vec<TraceEvent>>,
+	) -> Self {
+		let next_id = AtomicU64::new(1);
+		let mut targets: Vec<_> = targets
```
Should we instead convert the targets into a `Vec` in `BlockExecutor` and then pass them into here as a vec? That way in `patch_and_filter` we do not have to `split(",")` the original targets string again .. a tiny bit less code

insipx commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,294 @@
+			.map(|(_, s)| s.into())
+			.collect::<Vec<SpanDatum>>();
+
+		// Is there a way to do this sort without collecting twice? My only idea was to add start_time
```
Try using SpanDatum::from instead of the 'into()' method above. This should let rust recognize the type without collecting

emostov replied:
Don't we still have to `collect()` bc we need a `Vec` if we want to sort elements?

insipx commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,294 @@
+			.filter_map(|s| patch_and_filter(s, targets))
+			.collect();
+
+		// let events: Vec<_> = block_subscriber.events.lock().drain(..).map(|s| s.into()).collect();
```
Dead code (probably used for ref just putting this here to make sure it's removed later on)

insipx commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,294 @@
+}
+
+fn event_key_filter(event: &TraceEvent, storage_keys: &str) -> bool {
+	if let Some(key) = event.values.string_values.get("key") {
```
Maybe something like 

```rust
event.values
		.string_values.get("key")
		.and_then(|key| check_target(storage_keys, key, &event.level))
.unwrap_or(false)
```

I think it's more readable IMO

mattrutherford commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,294 @@
+
+		// Is there a way to do this sort without collecting twice? My only idea was to add start_time
+		// into `Span` - zeke
+		span_datums.sort_by(|a, b| a.start_time.cmp(&b.start_time));
```
I think we probably know the order by the span id, so perhaps we can save a little cost in the client by not sorting?

emostov replied:
@mattrutherford  I can easily sort client side by Id - but are Ids guaranteed to be ordered by time entered?

mattrutherford commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,294 @@
+
+		let block_subscriber = dispatch.downcast_ref::<BlockSubscriber>()
+			.ok_or("Cannot downcast Dispatch to BlockSubscriber after tracing block")?;
+		let mut span_datums: Vec<SpanDatum> = block_subscriber.spans
```
```suggestion
		let mut span_data: Vec<SpanDatum> = block_subscriber.spans
```

emostov commented:
### Response element sizes

`system::account` put/get event: 385 bytes
`:extrinsic_index` put/get event: 140 bytes
baseline response object: 525 bytes
span: 90 bytes

Note: event size can vary widely since they include the scale encoded data that is get/put into storage.

mattrutherford commented:
> ### Response element sizes
> 
> `system::account` put/get event: 385 bytes
> `:extrinsic_index` put/get event: 140 bytes
> baseline response object: 525 bytes
> span: 90 bytes
> 
> Note: event size can vary widely since they include the scale encoded data that is get/put into storage.
> ### Limiting max response size based off element sizes
> 
> In the default response with just have `system::account` and `:extrinsic_index` events. However since the user can specify all keys, the events could potentially be much bigger. My proposal is to assume the average event is 600 bytes (rough %50 bigger) and the average span is 100 bytes.
> The check for payload size would then look like:
> 
> ```rust
> 525_BYTES + events.len() * 600_BYTES + spans.len() * 100_BYTES < sc_rpc_server::MAX_PAYLOAD
> ```
> 
> This is certainly an non-exact heuristic, however short of an analysis of a decent number of blocks and events it seems like a good start and will work for keeping payloads well under the limit for the default use case.

Sounds better than a silent failure, if that's what happens currently when `sc_rpc_server::MAX_PAYLOAD` is exceeded. Perhaps as a follow-up PR it would be good to see if it's possible to implement an error response that could be applied to all RPCs in case this limit is reached.

emostov commented on primitives/rpc/src/tracing.rs:
```diff
@@ -0,0 +1,86 @@
+	pub string_values: HashMap<String, String>,
+}
+
+// fn default_level() -> Level {
```
dead code

emostov commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,317 @@
+			line: attrs.metadata().line().unwrap_or(0),
+			start_time: Instant::now(),
+			values,
+			overall_time: Default::default()
```
@mattrutherford - is this getting used anymore?

mattrutherford replied:
Some of these fields are used in other places in Substrate, but ultimately not in BlockSubscriber

emostov commented on primitives/state-machine/src/ext.rs:
```diff
@@ -201,11 +201,15 @@ where
-			HexDisplay::from(&key),
-			result.as_ref().map(HexDisplay::from)
+
+		let ready_to_encode_result = result.as_ref().map(|v| EncodeOpaqueValue(v.clone()));
```
Does this incur any additional computational cost when not tracing?

emostov commented on primitives/state-machine/src/ext.rs:
```diff
@@ -354,17 +327,21 @@ where
 			return;
 		}
 
+		let ready_to_encode_result = value.as_ref().map(|v| EncodeOpaqueValue(v.clone()));
```
Same question as above - does this incur a cost when not tracing?

niklasad1 replied:
yeah, I think so move into the the `trace! block` and it will only included when tracing/trace is enabled AFAIU.

emostov commented on primitives/state-machine/src/overlayed_changes/changeset.rs:
```diff
@@ -75,7 +75,6 @@ struct InnerValue<V> {
 #[derive(Debug, Clone)]
 #[cfg_attr(test, derive(PartialEq))]
 pub struct OverlayedEntry<V> {
-	/// The individual versions of that value.
```
I don't know whats going on here - is this intentional?

bkchr replied:
Please revert.

emostov commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,314 @@
+	}
+
+	fn new_span(&self, attrs: &Attributes<'_>) -> Id {
+		let id = self.next_id.fetch_add(1, Ordering::Relaxed);
```
@mattrutherford Just read the `Atomics` docs for the first time so bear with me ... with `Ordering::Relaxed` it does not seem like Ids are guaranteed to be in the order the span is entered. Can we change this too `Ordering::AcqRel`? 
From what I understand it should be minimal added cost on x86 platforms (bc x86 is strongly ordered) and then we can remove the sort and double collect for spans since we should be guaranteed that Id is incremented in the order `new_span` is invoked.

niklasad1 replied:
It's fine to "unordered" it just used as a simple counter and the ordered should not matter, Basti explained this very nicely [here](https://github.com/paritytech/substrate/pull/7780#discussion_r620558709)

However, as we talked about earlier it's easy to fuck it up if we decide that change that during refactoring or something. So better to wrap the `AtomicU64` is a newtype that only allows incrementing and not allow any other operations on it.

emostov replied:
> can you point me to "the sort and double collect for spans"? (Also, this comment was resolved, what was the outcome?)

This got taken out, since as long as we can get the guarantee that Id sequence corresponds to the sequence spans are entered then we don't need to include `time_entered`. Originally I was sorting on `time_entered`, which is not in the version of the span that gets sent over the wire. So I had to collect the spans from the subscriber, sort them and then map over them again to convert them to the type we send over the wire. However, no I just sort the spans by Id with api-sidecar.

emostov replied:
Let me know what to do with the ordering

emostov commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,314 @@
+		let id = self.next_id.fetch_add(1, Ordering::Relaxed);
+		let id = Id::from_u64(id);
+		let mut values = Values::default();
+		attrs.record(&mut values);
```
dq: why do we need to record `values` in `attrs` if its in the span?

mattrutherford replied:
Normally because spans can potentially have Fields too with values, however I think that we ignore that elsewhere, so this can probably be skipped for now.

emostov commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,314 @@
+
+	fn event(&self, event: &tracing::Event<'_>) {
+		let mut values = crate::Values::default();
+		event.record(&mut values);
```
Same q as above with `attrs.record` in `new_span`

mattrutherford replied:
Yes, this one we need because it's where all your storage keys etc are recorded

emostov commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,314 @@
+
+		// Is there a way to do this sort without collecting twice? My only idea was to add start_time
+		// into `Span` - zeke
+		span_data.sort_by(|a, b| a.start_time.cmp(&b.start_time));
```
Is there a way to do this sort without collecting twice? My only idea was to add `start_time` into `Span`

emostov commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,314 @@
+		if let Some(t) = span.values.string_values.remove(WASM_TARGET_KEY) {
+			span.target = t;
+		}
+		if !check_target(targets, &span.target, &span.level) {
```
Can we do this check first so we can exit early?

mattrutherford replied:
I don't think that would be any benefit because we still need to get the name and target from the values, which is all that's happening above really

dvdplm replied:
Lines 313-315 could be moved below this check.

emostov commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,308 @@
+			.drain()
+			.map(|(_, s)| SpanDatum::from(s))
+			.collect();
+		span_data.sort_by(|a, b| a.start_time.cmp(&b.start_time));
```
Is there a way to do this sort without collecting twice? My only idea was to add start_time into Span. (Moving comment out of code.)

emostov commented on primitives/rpc/src/tracing.rs:
```diff
@@ -0,0 +1,98 @@
+    pub error: String,
+}
+
+/// ``state_traceBlock` RPC response
```
```suggestion
/// `state_traceBlock` RPC response
```

emostov commented on client/rpc-api/src/state/mod.rs:
```diff
@@ -136,4 +136,98 @@ pub trait StateApi<Hash> {
+	///
+	/// [1]: https://crates.io/crates/tracing
+	/// [2]:
+	/// https://github.com/paritytech/substrate/tree/master/client/rpc-api/TraceBlock_Trace_Hierarchy.png
```
I think this link should work once the PR is merged

emostov replied:
Not sure why it won't show the PNG in this diff, but for now the link is here: 
 https://docs.google.com/drawings/d/1vZoJo9jaXlz0LmrdTOgHck9_1LsfuQPRmTr-5g1tOis/edit?usp=sharing

emostov commented:
@mattrutherford, @dvdplm, @insipx (and maybe @niklasad1 ?) - as far as my end is concerned everything looks good, ready for a review.

insipx commented on client/service/src/client/call_executor.rs:
```diff
@@ -81,15 +83,23 @@ where
+			debug!(target: "wasm_overrides", "using WASM override for block {}", id);
+			d
+		} else {
+			debug!(target: "wasm_overrides",
```
I'm not sure how useful this debug! line is, most of the time overrides won't be used, so this  will just be taking up space. 

IMO the line that tells us we are using an override is enough, if that line isn't printed we can assume that the override was not used

insipx reviewed (approved):
All LGTM here :+1:

bkchr commented on primitives/state-machine/src/ext.rs:
```diff
@@ -375,12 +360,6 @@ where
 		key: StorageKey,
 		value: Option<StorageValue>,
 	) {
-		trace!(target: "state", "{:04x}: PutChild({}) {}={:?}",
```
Why are these removed?

emostov replied:
They are not necessary for our goals with balance reconciliation and we wanted to reduce payload size.

bkchr replied:
All these are required for debugging. You could for example only change the level of the two you are interested in to debug?

Or maybe add some special tag (or whatever these extra arguments to the tracing log macros are called), so you can filter them on your own?

emostov replied:
I brought back the old traces and changed the target for the traces we need from `state` => `state_get_put`, this way if someone wants all the trace events in this file they can use `state` as the target and since it matches on prefix it will gather events on both targets.

cc @mattrutherford

mattrutherford replied:
Seems like the most simple soluition. The alternative would be to change the filtering to be more fine grained, by incorporating name and fields as well as target, as is used by tracing_subscriber EnvFilter:

https://docs.rs/tracing-subscriber/0.2.17/tracing_subscriber/filter/struct.EnvFilter.html

If we wanted to be compatible with that, then the syntax to match would I think be something like `state{Method=\"get\"}` etc...  which could be passed as the RPC params or set as a default

dvdplm replied:
I would prefer it if we made the filtering more fine grained as @mattrutherford suggests. Relying on the target string is a bit brittle.

emostov replied:
I tried out directive filtering in commit `6076975` with the expression `state[{method}]`; this correctly filtered the events but it also filtered out all the spans. Since we already have custom prefix based filtering on targets (i.e `metadata.target().starts_with(target`)I did not want to mess around with creating more directives based on the user supplied targets.

Instead I added a check on events our `Subscriber::enabled` impl.
Will comment there to point it out.
I am not convinced though that this is more stable since we still have rely on a particular field existing which seems as brittle as target; and in the current setup I have it gives the users less options.

cc @dvdplm @mattrutherford

bkchr commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,308 @@
+// Default to only pallet, frame support and state related traces
+const DEFAULT_TARGETS: &'static str = "pallet,frame,state";
+const TRACE_TARGET: &'static str = "block_trace";
+// Default to only events with the key prefixes for :extrinsic_index & system::Account.
+const DEFAULT_STORAGE_KEYS: &'static str =
+	"3a65787472696e7369635f696e646578,26aa394eea5630e07c48ae0c9558cef7b99d880ec681799c0cf30e8886371da9";
```
Putting FRAME related keys into the client isn't such a great idea.

emostov replied:
I can take these out. But would still need to keep the storage keys parameter so we can try and control payload size.
@dvdplm you think its fine to take these defaults out and then just include them in sidecar and maybe as an example in the docs?

bkchr replied:
Taking them as a parameter it complelty fine 👍

emostov replied:
Change the default to `""`, which means "don't filter out by storage keys". Added an example of storage key selection to the docs as well.

emostov commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,314 @@
+			if metadata.target().starts_with(target.as_str()) && metadata.level() <= level {
+				if metadata.is_span() {
+					return true;
+				} else if metadata.fields().field(REQUIRED_EVENT_FIELD).is_some() {
+					// it is an event, so we check if has the method field
+					return true;
+				}
```
This is the more fine grained filter for events.
@mattrutherford & @dvdplm let me know what you think.

bkchr commented on client/service/src/client/wasm_override.rs:
```diff
@@ -161,7 +161,13 @@ where
 				Some("wasm") => {
 					let wasm = WasmBlob::new(fs::read(&path).map_err(handle_err)?);
 					let version = Self::runtime_version(executor, &wasm, Some(128))?;
+					log::info!(target: "wasm_overrides",
+							   "Found wasm override in file: `{:?}`, version: {}",
+							   path.to_str(), version);
```
```suggestion
					log::info!(
						target: "wasm_overrides",
						"Found wasm override in file: `{:?}`, version: {}",
						path.to_str(),
						version,
					);
```

Same goes for all your log macro calls.

niklasad1 commented on client/rpc-api/src/state/mod.rs:
```diff
@@ -136,4 +136,106 @@ pub trait StateApi<Hash> {
+	/// ## Creating tracing enabled WASM runtimes
+	///
+	/// - Checkout commit of chain version to compile with WASM traces
+	/// - [diener](https://crates.io/crates/diener) can help to peg commit of substrate
```
why is this needed? Should be merged into master then you shouldn't need to use diener?

emostov replied:
The user will need to either create or get ahold of tracing enabled wasm binaries since by default they are not tracing enabled. These instructions explain how to create tracing enabled wasm binaries - which I don't think we would ever check into source control for substrate.

niklasad1 commented on client/service/src/client/call_executor.rs:
```diff
@@ -81,15 +83,23 @@ where
 			.transpose()?
-			.flatten()
-			.unwrap_or(onchain_code);
+			.flatten() {
```
I find this piece of code unnecessary complicated to read,

I would prefer to do it like this:

```rust
		let spec = self.runtime_version(id)?.spec_version;
		let wasm_override_code = self.wasm_override
			.as_ref()
			.and_then(|o| o.get(&spec, onchain_code.heap_pages));

		if let Some(code) = wasm_override_code = {
			debug!(target: "wasm_overrides", "using WASM override for block {}", id);
			Ok(code)
		} else {
			debug!(target: "wasm_overrides",
				"No WASM override available for block {}, using onchain code",
				id
			);
			Ok(onchain_code)
		}

``

insipx replied:
iirc `runtime_version` is in the `map` because calls to `runtime_version` can be expensive (requires executing a function in the WASM, and getting some values from the database). So to reduce unnecessary checking it was called conditionally if the overrides were present.

This is definitely more readable though; maybe the extra cost is worth it

niklasad1 replied:
ah, I see makes sense.

niklasad1 commented on client/service/src/client/call_executor.rs:
```diff
@@ -81,15 +83,23 @@ where
+		let code = if let Some(d) = self.wasm_override
 			.as_ref()
 			.map::<sp_blockchain::Result<Option<RuntimeCode>>, _>(|o| {
 				let spec = self.runtime_version(id)?.spec_version;
```
dq: is it intentional to return `Err` if runtime version couldn't be found?

insipx replied:
Yes, otherwise we can't know what runtime to use to execute that block. I suppose we could also coerce it to an Option but that will inevitably lead to silent failures that will be difficult to debug I think

bkchr replied:
Yes this is done by intention. Substrate runtimes need to provide a version. This is a fundamental requirement of substrate.

niklasad1 commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,314 @@
+// include the RPC result.
+const BASE_RPC_PAYLOAD: usize = 500;
+// Default to only pallet, frame support and state related traces
+const DEFAULT_TARGETS: &'static str = "pallet,frame,state";
```
```suggestion
const DEFAULT_TARGETS: &str = "pallet,frame,state";
```

niklasad1 commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,314 @@
+const DEFAULT_TARGETS: &'static str = "pallet,frame,state";
+const TRACE_TARGET: &'static str = "block_trace";
+// Default to not filtering based on storage keys as storage keys vary per chain.
+const DEFAULT_STORAGE_KEYS: &'static str = "";
```
```suggestion
const DEFAULT_STORAGE_KEYS: &str = "";
```

niklasad1 commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,314 @@
+// Default to not filtering based on storage keys as storage keys vary per chain.
+const DEFAULT_STORAGE_KEYS: &'static str = "";
+// The name of a field required for all events.
+const REQUIRED_EVENT_FIELD: &'static str  = "method";
```
```suggestion
const REQUIRED_EVENT_FIELD: &str  = "method";
```

niklasad1 commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,314 @@
+	fn enabled(&self, metadata: &tracing::Metadata<'_>) -> bool {
+		for (target, level) in &self.targets {
+			if metadata.target().starts_with(target.as_str()) && metadata.level() <= level {
+				if metadata.is_span() {
```
any particular reason why you are not doing:

```rust
if metadata.is_span || metadata.fields().field(REQUIRED_EVENT_FIELD).is_some() {
  return true;
}
```

emostov replied:
No reason, will change

niklasad1 commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,314 @@
+	}
+
+	fn new_span(&self, attrs: &Attributes<'_>) -> Id {
+		let id = self.next_id.fetch_add(1, Ordering::SeqCst);
```
I think this could be `Ordering::Relaxed` but it's doesn't matter that much because it's only faster on weakly-ordered platforms such as ARM

emostov replied:
We actually want the Ids to correspond to the order the spans are entered. 
My concern with using anything but the strictest ordering is that this assumption could break, although I have a poor understanding of async memory access so I could be wrong.

niklasad1 replied:
The id still atomic unless your using the ID for other synchronization such as indexing into memory or something similar.

But if you are not sure then you did the right call to make is `SeqCst` :)

bkchr replied:
I don't see any reason here as well to not use Relaxed.

Atomic ordering just instructs the compiler to insert barriers, that certain operations are done in a particular order. However, as here no ordering is required, relaxed should be enough.

And this also should be faster on x86_64, because they also support parallel execution of op codes.

emostov replied:
> I don't see any reason here as well to not use Relaxed.

@bkchr Would there be an issue if this function was called by a few separate threads at the same time? From what you are saying it sounds like not matter what atomic ordering is chosen it would not have an affect

bkchr replied:
You need certain ordering of operations for stuff where something relies on the fact that a certain variable is set before you lock/unlock. However in your case here you don't require such ordering. You only lock to insert into the hashmap and at this point the operation that has incremented the Atomic variable was already executed because it is required here.

emostov replied:
Makes sense, thanks for the explanation :)

niklasad1 commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,314 @@
+			name: event.metadata().name().to_owned(),
+			target: event.metadata().target().to_owned(),
+			level: *event.metadata().level(),
+			values: values.into(),
```
```suggestion
			values,
```

niklasad1 commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,314 @@
+		let id = BlockId::Hash(self.block);
+		let mut header = self.client.header(id)
+			.map_err(|e| format!("Invalid block Id: {:?}", e))?
+			.ok_or("Header not found".to_string())?;
```
```suggestion
			.ok_or_else(|| "Header not found".to_string())?;
```

niklasad1 commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,314 @@
+			.ok_or("Header not found".to_string())?;
+		let extrinsics = self.client.block_body(&id)
+			.map_err(|e| format!("Invalid block Id: {:?}", e))?
+			.ok_or("Extrinsics not found".to_string())?;
```
```suggestion
			.ok_or_else(|| "Extrinsics not found".to_string())?;
```

niklasad1 commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,314 @@
+	/// Execute block, recording all spans and events belonging to `Self::targets`
+	/// and filter out events which do not have the keys starting with one of the
+	/// prefixes in `Self::storage_keys`.
+	pub fn trace_block(&self) -> Result<TraceBlockResponse, String> {
```
why string as error type?

emostov replied:
@mattrutherford ?

niklasad1 commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,314 @@
+			.map_err(|e| format!("Invalid block Id: {:?}", e))?
+			.ok_or("Extrinsics not found".to_string())?;
+		tracing::debug!(target: "state_tracing", "Found {} extrinsics", extrinsics.len());
+		let parent_hash = header.parent_hash().clone();
```
H256 is copy.

```suggestion
		let parent_hash = *header.parent_hash();
```

niklasad1 commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,314 @@
+			.ok_or("Extrinsics not found".to_string())?;
+		tracing::debug!(target: "state_tracing", "Found {} extrinsics", extrinsics.len());
+		let parent_hash = header.parent_hash().clone();
+		let parent_id = BlockId::Hash(parent_hash.clone());
```
```suggestion
		let parent_id = BlockId::Hash(*parent_hash.clone);
```

niklasad1 commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,314 @@
+		let spans: Vec<_> = block_subscriber.spans
+			.lock()
+			.drain()
+			.map(|(_, s)| SpanDatum::from(s))
```
this looks redundant

niklasad1 commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,314 @@
+// Estimate of the max size base RPC payload when the Id is bound as a u64. If strings
+// are used for the RPC Id this may need to be adjusted. Note: The base payload does not
+// include the RPC result.
+const BASE_RPC_PAYLOAD: usize = 500;
```
You may want the clarify that we estimate the max `base size` of the JSONRPC response message and it has the following format:

`{"jsonrpc":"2.0","result":[],"id":18446744073709551615}`

If we ignore huge strings as IDs it can only be 55 bytes, https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=85558e03fb6f28dcf36484d7f9b6714c but 500 is probably fair ^^

emostov replied:
>  https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=85558e03fb6f28dcf36484d7f9b6714c
This is a cool example!

niklasad1 commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,314 @@
+		.unwrap_or(false)
+}
+
+fn patch_and_filter(mut span: SpanDatum, targets: &str) -> Option<Span> {
```
Can we document this function? It's not obvious to me what this does and what it's side-effects has for impact.

Such as: 
if the `WASM_NAME_KEY` is removed the span.name is set to that
if the `WASM_TARGET_NAME` is removed the span.target is set to that

emostov replied:
@mattrutherford can you put a comment on this to explain some background on the the special wasm targets.

emostov replied:
btw I put in a comment above this

niklasad1 commented on primitives/rpc/src/tracing.rs:
```diff
@@ -0,0 +1,98 @@
+#[serde(rename_all = "camelCase")]
+pub struct BlockTrace {
+	/// Hash of the block being traced
+	pub block_hash: String,
```
why not H256?

emostov replied:
Won't it need to be allocated anyways since we always serialize?

niklasad1 replied:
yeah it would allocate when serializing/deserializing but current we do this:

```rust
		let block_trace = BlockTrace {
			block_hash: block_id_as_string(id),
			parent_hash: block_id_as_string(parent_id),
			tracing_targets: targets.to_string(),
			storage_keys: storage_keys.to_string(),
			spans,
			events,
		};

```

where block_id is to H256 which already implements serde AFAIK so it unnecessary to convert it a String and then serialize it :)

emostov replied:
`block_id_as_string` has a special way of displaying that is a bit easier to read for the RPC caller:
```
fn block_id_as_string<T: BlockT>(block_id: BlockId<T>) -> String {
	 match block_id {
		BlockId::Hash(h) => HexDisplay::from(&h.encode()).to_string(),
		BlockId::Number(n) =>  HexDisplay::from(&n.encode()).to_string()
	}
}
```
So I think we would need to write a custom `Serialize` impl that called `block_id_as_string` .. I might give it a try if I have time.

emostov commented on primitives/rpc/src/tracing.rs:
```diff
@@ -0,0 +1,98 @@
+	/// Hash of the block being traced
+	pub block_hash: String,
+	/// Parent hash
+	pub parent_hash: String,
```
We serialize it in a specific way prior to putting it into this response struct. I suppose we could do the same in a custom `Serialize` impl and then use H256 here ... but not worth changning IMO

niklasad1 replied:
your call :)

dvdplm commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,329 @@
+			spans,
+			events,
+		};
+		let block_trace_size = serde_json::to_vec(&block_trace)
+			.map_err(|e| format!("Failed to serialize payload: {}", e))?
+			.len();
```
This is a bit problematic, we serialize here and then once again before returning? I don't have a better idea atm but will try to come up with something.

I guess we mostly care about the `spans` and `events` members so perhaps we can keep a running count of them during execution? Or how about: we make a guesstimate of the average size of a `span` and `event` and then use that heuristic to calculate `approx_size_of_a_span * block_traces.spans.len()` (and same for events); I don't think we need to be very precise.

We could also move this check to where we return the JSON: serialize, check for length and if too big, return an error?

niklasad1 replied:
yeah, that's a good idea I guess we could serialize the response to a String [here](https://github.com/paritytech/substrate/pull/7780/files#diff-d155229ffb6c782e34d8874b260cefb3b90d2b8549661b40c5b469f0fe125e0bR537) instead re-use it.

emostov replied:
> we make a guesstimate of the average size of a span and event and then use that heuristic to calculate approx_size_of_a_span * block_traces.spans.len() (and same for events); I don't think we need to be very precise.

If we want to go this route, I did some work to get estimates before [here](https://github.com/paritytech/substrate/pull/7780#issuecomment-819685718)

emostov replied:
Ok I just changed it to

```rust
		let aprox_payload_size = BASE_PAYLOAD + events.len() * AVG_EVENT + spans.len() * AVG_SPAN;
		let response = if aprox_payload_size > MAX_PAYLOAD {
				TraceBlockResponse::TraceError(TraceError { ... })
		} else {
			TraceBlockResponse::BlockTrace(BlockTrace { ... })
		};
```

emostov replied:
Ok i just changed it back to:

```rust
		let aprox_payload_size = BASE_PAYLOAD + events.len() * AVG_EVENT + spans.len() * AVG_SPAN;
		let response = if aprox_payload_size > MAX_PAYLOAD {
				TraceBlockResponse::TraceError(TraceError { ... })
		} else {
			TraceBlockResponse::BlockTrace(BlockTrace { ... })
		};
```

dvdplm replied:
> yeah, that's a good idea I guess we could serialize the response [here](https://github.com/paritytech/substrate/pull/7780/files#diff-d155229ffb6c782e34d8874b260cefb3b90d2b8549661b40c5b469f0fe125e0bR537) instead but not sure if it's possible to construct a `JsonRpcResponse` without double allocate anyway.

Yeah it looks a bit messy to get right. I say we punt on it and make do with the estimate Zeke implemented.

emostov commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,329 @@
+
+impl Subscriber for BlockSubscriber {
+	fn enabled(&self, metadata: &tracing::Metadata<'_>) -> bool {
+		for (target, level) in &self.targets {
```
I believe this is called for every span single span and event that the subscriber encounters (so a lot).

Yeah we need to match on partial prefix. I honestly don't have any great ideas unless we wanted to use some sort of trie that accommodates prefix matching. 

In the default scenario, targets is a constant of 3; and in other expected use cases I Imagine targets will not get too long (e.g. how often are people specifying over 10). 

Do you have ideas for how to go about benchmarking this? And considering we are still trying to ship an MVP do you think its worth doing?

dvdplm replied:
My idea wouldn't work anyway. It's not possible to use `(String, Level)` as a hash key because of [this](https://github.com/tokio-rs/tracing/issues/1376). Linear search on such a small will have to due for now.

emostov replied:
I don't think we should do it but we could use 2 data structures:

1) HashMap `target => level`
2) Prefix trie with targets

So first see if the event/span target is in the prefix trie. If it is a match, take the positive target from prefix trie and then key into the HasMap to get the level.

dvdplm replied:
Worth pondering but I agree we leave be for now.

emostov commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,329 @@
+	fn enabled(&self, metadata: &tracing::Metadata<'_>) -> bool {
+		for (target, level) in &self.targets {
+			if metadata.target().starts_with(target.as_str()) && metadata.level() <= level {
+				if metadata.is_span() || metadata.fields().field(REQUIRED_EVENT_FIELD).is_some() {
```
Yup, good point

dvdplm commented on client/rpc-api/src/state/mod.rs:
```diff
@@ -136,4 +136,103 @@ pub trait StateApi<Hash> {
+	///
+	/// ## Node requirements
+	///
+	/// - Fully synced archive node.
```
What happens if you run this on a partially synced node, i.e. you call the RPC on a block that is present in the database (but we're still syncing). I would have expected that to work.
And the same question for non-archive nodes, again I would have expected the re-execution to work just as long as the block is actually available (i.e. hasn't been pruned).

emostov replied:
Just from trial:
-  a non-archive node seems to not keep any block bodies around so it fails quickly
- syncing nodes seem to have undefined behavior - most of the time it works but sometimes it doesn't. From my exprience of supporting api-sidecar I have seen borked data from queries to a actively syncing node

When I say syncing I mean a "major" sync, as in its trying to catch up on a large amount of blocks; not just actively syncing the latest blocks (which every node should be doing).

dvdplm replied:
Ok, good to know. A non-archive node should keep the most recent 256 blocks around though so that's a bit odd.

dvdplm commented on client/rpc-api/src/state/mod.rs:
```diff
@@ -136,4 +136,103 @@ pub trait StateApi<Hash> {
+	/// The filtering of spans and events takes place after they are all collected; so while filters
+	/// do not reduce time for actual block re-execution, they reduce the response payload size.
+	///
+	/// Note: storage events primarily come from _primitives/state-machine/src/ext.rs_).
```
That `)` looks funny, should it be removed?

dvdplm commented on client/rpc-api/src/state/mod.rs:
```diff
@@ -136,4 +136,103 @@ pub trait StateApi<Hash> {
+	/// 	(no `0x` prefix) storage keys. If an empty string is specified no events will
+	/// 	be filtered out. If anything other than an empty string is specified, events
+	/// 	will be filtered by storage key (so non-storage events will **not** show up).
+	/// 	You can specify any degree of a storage key prefix (i.e. if a specified storage
```
"degree"? You mean length maybe?

dvdplm commented on client/rpc-api/src/state/mod.rs:
```diff
@@ -136,4 +136,103 @@ pub trait StateApi<Hash> {
+	/// narrow down the traces using a smaller set of targets and/or storage keys.
+	///
+	/// If you are having issues with maximum payload size you can use the flag
+	/// `-lstate_tracing=trace` to get some logging during tracing.
```
This is probably the most thoroughly documented RPC in all of substrate, great job.

Maybe add an example `curl` command?
Maybe add a link to the `substrate-archive` wasm blobs collection (as I suspect that is the trickiest step when setting this up)

emostov commented on client/tracing/src/block/mod.rs:
```diff
@@ -300,15 +298,15 @@ fn event_key_filter(event: &TraceEvent, storage_keys: &str) -> bool {
+// is for substrate's WASM tracing wrappers to put the `name` and `target` data in the `values` map
+// (normally they would be in the static metadata assembled at compile time). Here, if a special
+// WASM `name` or `target` key is found in the `values` we remove it and put the key value pair in
+// the span's metadata, making it consistent with spans that come from native code.
```
I found this really helpful. @dvdplm Do you think it would make sense to add a modified version of this comment directly to the WASM tracing code as well.

dvdplm replied:
I think it bears repeating for sure. Do you know of a specific place to put it? Please do!

emostov replied:
See https://github.com/paritytech/substrate/pull/7780/commits/6e3c3366fbb9137c6391cffa5b8517a4a0ace75b

If anyone finds that controversial feel free to revert

emostov commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,341 @@
+	MissingBlockComponent(String),
+	#[error("Dispatch error: {0}")]
+	Dispatch(String)
+}
```
@dvdplm or @niklasad1 Does this look ok for errors? (As opposed to the plain `String` errors we used to have)

dvdplm replied:
Works.

I changed the name of the `Result` type alias to `TraceBlockResult` for clarity.

dvdplm reviewed (approved):
LGTM.

dvdplm commented on client/tracing/src/logging/mod.rs:
```diff
@@ -171,6 +171,9 @@ where
 	let builder = FmtSubscriber::builder().with_env_filter(env_filter);
 
+	#[cfg(not(target_os = "unknown"))]
+	let builder = builder.with_span_events(format::FmtSpan::CLOSE);
```
@emostov this change causes [this test](https://github.com/paritytech/substrate/blob/master/client/tracing/src/logging/mod.rs#L387) to fail. Changing it to `FmtSpan::NONE` which seems to work but I honestly don't know what I'm doing here. :/

emostov replied:
See https://github.com/paritytech/substrate/pull/7780/commits/6e3c3366fbb9137c6391cffa5b8517a4a0ace75b

If anyone finds that controversial feel free to revert

dvdplm commented:
@bkchr any objections on your end to merging this?

bkchr commented on primitives/state-machine/src/ext.rs:
```diff
@@ -354,17 +363,26 @@ where
+			method = "Put",
+			ext_id = self.id,
+			key = %HexDisplay::from(&key),
+			result = %HexDisplay::from(&value.encode()),
+			result = %HexDisplay::from(
+				&value
+					.as_ref()
+					.map(|v| EncodeOpaqueValue(v.clone()))
+					.encode()
+			),
```
```suggestion
			value = %value.as_ref().map(HexDisplay::from),
			value_encoded = %HexDisplay::from(
				&value
					.as_ref()
					.map(|v| EncodeOpaqueValue(v.clone()))
					.encode()
			),
```

dvdplm replied:
Not sure if it's ok to rename the keys here, cc @emostov

emostov replied:
@bkchr & @dvdplm 

I changed both to look like:

```rust
		trace!(
			target: "state",
			method = "Get",
			ext_id = self.id,
			key = %HexDisplay::from(&key),
			result = ?result.as_ref().map(HexDisplay::from),
			result_encoded = %HexDisplay::from(
				&result
					.as_ref()
					.map(|v| EncodeOpaqueValue(v.clone()))
					.encode()
			),
		);
		```

emostov replied:
However my preference would be leave out the non-encoded field since I think it would only be useful for visual inspection (and still it will probably just be an array of ints, really sure if thats helpful).

emostov replied:
See commit `1f3ad4e` for changes

emostov replied:
Once this is settled though looks ready to merge ?

bkchr replied:
> However my preference would be leave out the non-encoded field since I think it would only be useful for visual inspection (and still it will probably just be an array of ints, really sure if thats helpful).

I have already used this quite often for debugging state mismatches :D And seeing `Some(hash)` or `None` is still better than parsing just the raw encoded data.

bkchr replied:
@emostov why not call it here `value =` and `value_encoded`? As this is not the result in this context.

emostov replied:
@bkchr just trying to avoid extra logic on the processing side, but I will change it to your suggestion so it is more readable in context.


> I have already used this quite often for debugging state mismatches :D And seeing Some(hash) or None is still better than parsing just the raw encoded data.

I needed to change
```
result = ?result.as_ref().map(HexDisplay::from),
```
to use the `?` operator, which is tracing shorthand for `Debug`, should that be ok for your purposes? (Using `%` for `Display` had some issues and I didn't try for to long to sort it out)
[see here for tracing field shorthand docs](https://docs.rs/tracing/0.1.26/tracing/index.html#recording-fields)

bkchr replied:
Yeah I know the tracing operators ;) 

It was actually a mistake by me to suggest `%`. Before we also used `Debug`, so everything fine now :)

bkchr commented on primitives/state-machine/src/ext.rs:
```diff
@@ -201,11 +201,20 @@ where
+			method = "Get",
+			ext_id = self.id,
+			key = %HexDisplay::from(&key),
+			result = %HexDisplay::from(
+				&result
+					.as_ref()
+					.map(|v| EncodeOpaqueValue(v.clone()))
+					.encode()
+			),
```
```suggestion
			result = %value.as_ref().map(HexDisplay::from),
			result_encoded = %HexDisplay::from(
				&result
					.as_ref()
					.map(|v| EncodeOpaqueValue(v.clone()))
					.encode()
			),
```

dvdplm replied:
Same here, cc @emostov

dvdplm replied:
I'll add comments about these key names so people don't change them without thinking.

bkchr reviewed (approved):
Some last "nitpicks" otherwise looks good so far

dvdplm commented on client/tracing/src/block/mod.rs:
```diff
@@ -0,0 +1,341 @@
+			s
+		} else {
+			DEFAULT_STORAGE_KEYS
+		};
```
I got rid of it altogether [here](https://github.com/paritytech/substrate/pull/7780/commits/b6f9dd19b212d45d8d33102f4443f72a073836fe#diff-f5ee9da45314c57567317ca337c1686d02bcf4d13779775ec99186d275e51957R262-R266) – @emostov can you double check I didn't fuck up the semantics there?

emostov replied:
This seems fine assuming default is an empty `String`.

xlc commented:
Any plan to build some debugging tools with this or if there are already something available?

emostov commented:
> Any plan to build some debugging tools with this or if there are already something available?

@xlc The goal for the tooling team was to build a non-event based balance reconciliation MVP with future goals of more generic state change tracking/auditing tooling. 
- https://github.com/paritytech/substrate-api-sidecar/pull/383
- https://github.com/paritytech/substrate-reconciler
And worth checking the docs for the RPC itself for some more details
